Correct fallbacks to empty prefix in ModelBinding (#1865 and #2129)

dougbu opened the pull request:
this PR is mostly complete but lacks unit and functional tests covering the specific fallback scenarios mentioned in #1865 and #2129.  i.e. I didn't have to update as many tests as I'd hoped.

the first three commits are cleanup I did as I was working out what the model binding system was doing.  might be easier to look at them individually or at least review all the commit descriptions before looking at the files.

dougbu commented:
/cc @harshgMSFT @rynowak

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/ModelBindingResult.cs:
```diff
@@ -23,23 +23,60 @@ public ModelBindingResult(object model, string key, bool isModelSet)
+        /// <c>false</c>.
+        /// </param>
+        public ModelBindingResult(object model, string key, bool isModelSet, bool isEmptyModel)
+            : this(model, key, isModelSet)
```
should this be the other way around? Typically a constructor with less parameters calls the one with more.

dougbu replied:
typically != required.  I chose to have the rarely-used constructor depend on the common one.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -34,17 +35,35 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
+                    fallback = true;
+
+                    // Remove any validation errors from first attempt; these are likely Required errors.
+                    // Clear entries completely, unlike ModelState.ClearValidationState().
```
wouldnt `ClearValidationState` clear all entries with the model name?  Is there a test which I can see to understand the scenario?

dougbu replied:
`ClearValidationState` doesn't remove the entries, it just clears the contained `ModelState.Entries` entries and removes `ModelState.ValidationState` references.  this leaves the `ModelState.Value` untouched.  that isn't correct for fallback scenarios.

I mentioned in the PR description that unit and functional tests are to come.  the bugs are:
- #1865 (still needs a reword): `MutableObjectModelBinder` returns a non-`null` result containing a value whenever it can create that value.  this looks to the calling `CompositeModelBinder` like success and prevents fallback to the empty prefix when no properties in a top-level model are actually set.
- #2129: `GenericModelBinder` returns a non-`null` result whenever it can create a binder for the given type.  this also prevents fallback to the empty prefix when no provider has data.

this particular code is part of the #1865 fix.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs:
```diff
@@ -399,11 +405,14 @@ internal void ProcessDto(ModelBindingContext bindingContext, ComplexModelDto dto
             }
 
             // for each property that was bound, call the setter, recording exceptions as necessary
+            bool isModelSet = false;
```
var

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -34,17 +35,35 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
-                                                            modelName: string.Empty);
-                modelBindingResult = await TryBind(newBindingContext);
+                var fallback = modelBindingResult == null;
+                if (!fallback && modelBindingResult.IsModelSet && modelBindingResult.IsEmptyModel)
```
Why is this interesting to do only for an empty model? Technically we should reset all properties which have an entry and have a default value? would like to understand more.

dougbu replied:
in the other fallback scenarios, the `ModelState` won't contain anything for the previous prefix

rynowak commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -34,17 +35,35 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
+                    foreach (var entry in entries.ToList())
+                    {
+                        bindingContext.ModelState.Remove(entry.Key);
+                    }
```
I'm concerned that this is needed. This code keeps getting more complex as we fix bugs, which is a hint that we haven't solved the design issues.

dougbu replied:
I suggest the underlying issue is less the design than the apparent requirement to keep filling in empty objects when nothing from the value providers match.  see for example the closed issue #1488.  `MutableObjectModelBinder` is thus going to be a strange beast regardless of the surrounding architecture.

rynowak commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs:
```diff
@@ -323,6 +323,12 @@ internal static PropertyValidationInfo GetPropertyValidationInfo(ModelBindingCon
+                {
+                    // Skip indexer properties and others ModelMetadata ignores.
+                    continue;
+                }
```
unrelated?

dougbu replied:
hit the NRE when I changed something else.  but I split it into a separate commit if we want to get it in separately.

harshgMSFT replied:
Would like to know what that case was ?

dougbu replied:
a POCO model that has an indexer property

rynowak replied:
This is a consequence of #1712. I think it's good to add the workaround here for now, I'll be fixing it this week.

rynowak commented on src/Microsoft.AspNet.Mvc.ModelBinding/ModelBindingResult.cs:
```diff
@@ -23,23 +23,60 @@ public ModelBindingResult(object model, string key, bool isModelSet)
+        /// properties.
+        /// </para>
+        /// </summary>
+        public bool IsEmptyModel { get; }
```
This really doesn't feel right. This seems intended to undo the logic in `MutableObjectModelBinder` which figures out  whether to try and bind a POCO model. If we don't have the data necessary there to make a decision then we need to re-evaluate the design, not work around it.

dougbu replied:
rinse and repeat https://github.com/aspnet/Mvc/pull/2188#discussion_r26549541

this isn't intended to undo the (strange) `MutableObjectModelBinder` behaviour but to inform `CompositeModelBinder` that one of those behaviours occurred.  in addition the `ModelBindingResult` is the only way binders "communicate".

nothing changes about regular POCO creation nor filling that object into a larger model.  the only use of this flag (currently) is for `CompositeModelBinder` to realize it should invoke its fallback-to-empty-prefix logic.

dougbu commented:
@rynowak @harshgMSFT rebased and updated.  changed mind after going back to earlier approach we discussed.  instead _removed_ a special-case that seemed to have a few other poor side effects.

dougbu commented:
should be done except for perhaps one or two unit tests and a functional test for top-level collections (#2129)

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -34,14 +33,17 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
-                && bindingContext.FallbackToEmptyPrefix)
+
+            if (modelBindingResult == null &&
+                bindingContext.FallbackToEmptyPrefix &&
```
Any specific reason to change the order ? I am guessing none.

dougbu replied:
mostly. I was mainly re-wrapping the condition because it looked strange but chose this order because it's `FallbackToEmptyPrefix` that makes `IsNullOrEmpty` relevant.

harshgMSFT replied:
:+1

dougbu commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -728,20 +728,21 @@ public class ModelBindingTest
+            Assert.Equal("The Field4 field is required.", json["Field4"]);
+            Assert.Equal("The Field5 field is required.", json["Field5"]);
+            Assert.Equal("The Field6 field is required.", json["Field6"]);
+            Assert.Equal("The Field7 field is required.", json["Field7"]);
```
fyi this change should not have been necessary.  also indicates a bug because the query provides values for the top-level properties.

the broken `MutableObjectModelBinder` must have been hiding problems in another binder. looking for the binder that returns `null` or a result with `!IsModelSet` when prefixes match in the value providers...

harshgMSFT replied:
Is this the result of the check removed in https://github.com/aspnet/Mvc/pull/2188/files#diff-62a5557f611407fe0ee3bd9e9e2a1393L96 ?

dougbu replied:
indirectly.  bugs in `MutableObjectModelBinder` were hiding bugs in other binders.  as I said, I'll find the offender.

dougbu replied:
I was wrong.  this test was demonstrating an unintended side-effect of `MutableObjectModelBinder` creating non-top-level POCO objects even when no data was available.  the correct thing is for required properties in the top-level object to be validated, as happens now.

dougbu commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -753,9 +754,9 @@ public class ModelBindingTest
-            Assert.Equal("The Field3 field is required.", json["model.Field3"]);
+            Assert.Equal("The Field1 field is required.", json["Field1"]);
+            Assert.Equal("The Field2 field is required.", json["Field2"]);
+            Assert.Equal("The Field3 field is required.", json["Field3"]);
```
this demonstrates the #1865 fix

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -34,14 +33,17 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
             {
-                // fallback to empty prefix?
+                // Remove any validation errors from first attempt; these are likely Required errors.
+                bindingContext.ModelState.Clear();
```
If I understand correctly, we fall in this case only when none of the model binders was chosen. Which means there should not have been any errors/ values in the model state dictionary. Is my understanding incorrect?

rynowak replied:
It's not safe to clear model state like this, because it could have validation errors from other parameters.

dougbu replied:
right.  this shouldn't be needed anymore.  just a holdover from my previous fixes.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs:
```diff
@@ -52,7 +51,13 @@ protected virtual bool CanUpdateProperty(ModelMetadata propertyMetadata)
+            // Reviewers: The isTopLevelObject value is wrong whenever a binder (e.g. CollectionModelBinder) uses
+            // GetMetadataForType(). Suggest we need ModelMetadata.GetMetadataForElementType() and a special case here
+            // for ComplexModelDto. Impact will likely be mainly extra "empty" (no bound property) objects created.
+            // Worth filing an issue?
```
I dont think we should do `ModelMetadata.GetMetadataForElementType()` mainly because this code is agnostic of what is being bound here ( collection vs not a collection). I would file an issue for this and move the design discussion there. 
Also lets yank the comment out ( or atleaset remove the suggestion).

dougbu replied:
it's a "Reviewers" comment. I have no intention of leaving it in.

harshgMSFT replied:
cool, Typically i have seen people commenting on PR itself to bring out something. I know typically != required :smiling_imp:

rynowak replied:
Sounds like we need to add `IsTopLevelObject` to the context instead then. Agree with @harshgMSFT - we don't want more model metadata APIs.

dougbu replied:
started writing up an issue and convinced myself our current binders don't use `GetMetadataForType()` enough to make this issue visible. might be worth adding `IsTopLevelObject` for protection going forward but:
- existing `isTopLevelObject` value is generally correct unless the model type is an element of a collection
- that incorrect value matters only when either the prefix is empty (fine) or the `BinderModelName` is not `null`
- the `IModelNameProvider` attributes can't be applied to a class

dougbu replied:
filed #2203 but mentioned it isn't urgent

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs:
```diff
@@ -90,16 +95,7 @@ protected virtual bool CanUpdateProperty(ModelMetadata propertyMetadata)
                 return true;
             }
 
-            // 3. The model name is not prefixed and a value provider can directly provide a value for the model name.
```
Why is this removed?

dougbu replied:
from commit comments:
> - change `MutableObjectModelBinder` to ignore exact match in value providers
 - had an incorrect assumption: don't want exact model name to match since
   this binder supports only complex objects
 - ignored `BinderModelName`, value provider filtering, et cetera

dougbu replied:
put another way, `bindingContext.ValueProvider.ContainsPrefixAsync()` _always_ returns `true` if an exact match exists so the commend and code are incorrect.  could probably make it almost work passing `bindingContext.ModelName + "."` but `CanValueBindAnyModelProperties()` is more careful and works better in some scenarios.

harshgMSFT replied:
Does this relate to the fix that we are looking at or is it a general cleanup? 
The comment is correct, it never said that ContainsPrefixAsync does not return true if an exact match exists. 
Complex objects can be directly bound by the value providers, unless there is a strong reason for removing this i would like to keep this. 
> works better in some scenarios.

Can you point me to those scenarios?

dougbu replied:
> this relate to the fix that we are looking at

it _is_ the fix

> Complex objects can be directly bound by the value providers

what value providers do this?  even if such exist or someone could write one, `MutableObjectModelBinder` can't handle the case -- it always creates `ComplexModelDto` and binds the individual properties.

> Can you point me to those scenarios?

cases involving `BinderModelName` mostly. look at the special cases in `CanValueBindAnyModelProperties` and `CanBindValue`

harshgMSFT replied:
Aa I see, 
so basically when we fallback on empty string, this check caused non top level model to get created even though there was nothing which was binding a value? I am sorry This is confusing, it would be great if you can walk me through an example.

dougbu replied:
close.  this case was hit for any `ModelName` that looked like a top-level parameter or (after the fallback to empty prefix) property name.  e.g. the #1865 scenario where top-level POCO model was (due to this code) created the first time through.  so `CompositeModelBinder` didn't fallback to the empty prefix.

rynowak commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -97,7 +100,14 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
                 if (result != null)
                 {
-                    return result;
+                    if (result.IsModelSet || bindingContext.ModelState.ContainsKey(bindingContext.ModelName))
```
This condition needs to docs as to the rationale. Not all model binders will add data to model state.

dougbu replied:
sure. I'll add a comment.

added this code a while ago but believe the reason was conversion errors.  if a binder found _any_ value, we don't want to fall back to the empty prefix.  could also change `TypeConverterModelBinder` to return a result with `IsModelSet == true` when it hits a conversion `Exception`. but the effect would be the same and might lead to other behaviour changes elsewhere. worth trying from your POV?

rynowak commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -1949,9 +1950,11 @@ public class ModelBindingTest
+
+            // Reviewers: Slightly odd behavior is specific to unusual error cases. Happens because
+            // MutableObjectModelBinder no longer incorrectly creates a model when value providers have no data.
             var address = Assert.Single(result.Addresses);
```
A POCO parameter should always get a value. Did this change regress that?

dougbu replied:
no.  `result` is non-`null` because the `UserWithAddress` parameter was bound in the controller.  however `PersonAddress` is a non-top-level POCO object that was previously bound even though no data existed for it.

sequence is something like:
- `MutableObjectModelBinder` sees data for its `Addresses` property exists
- `ComplexObjectDtoModelBinder` attempts to bind `Addresses`
- `CollectionModelBinder` starts with the `0` property because `bindingContext.ValueProvider.ContainsPrefixAsync()` returns `true` (a case it can normally handle)
- attempt to convert the raw value to a collection of `PersonAddress` fails so entry gets `null` value

harshgMSFT commented:
:watch:

dougbu commented:
@rynowak @harshgMSFT rebased and updated.  still needs one or two more `MutableObjectModelBinder` unit tests and a #2129 functional test.  but very close to non-[Design]

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -97,7 +97,17 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
                 {
-                    return result;
+                    // Use returned ModelBindingResult if it either indicates model was set or is related to errors.
+                    // The second condition is necessary because some binders add errors that would be duplicated when
```
This code is saying that return a successful result even if there was no model bound by the model binder. I am worried reading the comment, just to avoid duplicate errors we should not do this. Can you throw in more examples?

dougbu replied:
I'll generalize this comment because `TypeConverterModelBinder` and `TypeMatchModelBinder` both include calls to `bindingContext.ModelState.SetModelValue()`.  in `TypeConverterModelBinder`'s case, that may be associated with returning a result with `!IsModelSet`.

in any case this code is about whether to maintain the current result or `null`.  it will exit the loop regardless.

the difference between returning `null` and a result with `!IsModelSet` (the only case where the added `ContainsKey()` check matters) is that the caller may fall back to the empty prefix in the first case but will not in the second.

further extra entries in the `ModelState` will lead to `!IsValid` since validation won't change their `Unvalidated` state.

harshgMSFT replied:
What I was pointing out was, you should specifically check for validation state as Error and not anything else.

dougbu replied:
:-1: if there's any `ModelState`, either an error occurred or one of the binders successfully added a value but oddly returned a !IsModelSet value.

rynowak replied:
Not sure I have an opinion on this. If anything if we get scenarios where a model binder is buggy we'll at least know...

rynowak replied:
Wondering if this should throw? Should we be concerned about cases where keys overlap?

dougbu replied:
which "this"?  the case where `!IsModelSet && bindingContext.ModelState.ContainsKey(bindingContext.ModelName) && thatModelState.IsValid`?  all I know is it's freaking hard to remove the `bindingContext.ModelState.ContainsKey(bindingContext.ModelName)` bit and get the 10s of failing tests working again.  not sure if any are hitting that particular case.

and is that really part of this PR?  I didn't _add_ cases like that or even change anything in `TypeConverterModelBinder` or `TypeMatchModelBinder` except to name parameters.  we're talking about whether or not to return `null` instead of a slightly odd result that would have propagated up before I started on these bugs.

harshgMSFT replied:
Should we quickly sync and resolve this?

harshgMSFT commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -1951,9 +1952,12 @@ public class ModelBindingTest
             var result = await ReadValue<UserWithAddress>(response);
+
+            // Though posted content did not contain any Addresses, it is bound as a single-element List containing
+            // null.  Slightly odd behavior is specific to this unusual error case and fact Address is POCO.
```
nit: extra space before `Slightly`

harshgMSFT commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -1951,9 +1952,12 @@ public class ModelBindingTest
             // Assert
             var result = await ReadValue<UserWithAddress>(response);
+
+            // Though posted content did not contain any Addresses, it is bound as a single-element List containing
```
nit: `any valid Addresses`

harshgMSFT replied:
Why is this the behavior, should it be a list with no element?

rynowak replied:
This behavior seems like it's still a bug. @dougbu changes change it to be a little cleaner, but ideally this would be an empty list.

harshgMSFT replied:
do we want to track it separately?

harshgMSFT commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -1995,7 +1999,7 @@ public class ModelBindingTest
 
         [Fact]
-        public async Task BindModelAsync_WithNestedCollectionContainingRecursiveRelation_WithMalformedValue()
+        public async Task BindModelAsync_WithNestedCollectionContainingRecursiveRelation_WithMalformedValue_BindsSingleNullEntry()
```
nit: line break

harshgMSFT commented on test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs:
```diff
@@ -2012,9 +2016,12 @@ public class ModelBindingTest
             // Assert
             var result = await ReadValue<PeopleModel>(response);
+
+            // Though posted content did not contain any People, it is bound as a single-element List containing
```
same comments as above.

dougbu replied:
1. @harshgMSFT could you file the bug about `GenericModelBinder` and some / all (??) subclasses not creating the expected empty collection in some error cases?  it really is a corner of an error case because `new[] { null }` is actually the correct model when data exists for item `0` but that data is invalid / cannot be converted to the target type. I'd rather not spend further time describing what's going on here.
2. sure I'll remove the extra spaces

dougbu replied:
oh, and
3. sure I'll add the word "valid"

rynowak replied:
@dougbu - we already have this bug.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/CompositeModelBinder.cs:
```diff
@@ -97,7 +97,17 @@ public CompositeModelBinder([NotNull] IEnumerable<IModelBinder> modelBinders)
+                    // Use returned ModelBindingResult if it either indicates model was set or is related to errors.
+                    // The second condition is necessary because some binders add errors that would be duplicated when
+                    // falling back to the empty prefix. Those binders often cannot run twice anyhow.
+                    if (result.IsModelSet || bindingContext.ModelState.ContainsKey(bindingContext.ModelName))
```
Moreover this key might not have a validation state of an error, in which case you would want to continue.

dougbu replied:
this loop has always exited when it hit the first `binder` to return a non-`null` result.  all I've done is changed the method to return `null` in a few more cases, letting the caller fall back to the empty prefix.

harshgMSFT commented:
:watch:

rynowak commented on src/Microsoft.AspNet.Mvc.ModelBinding/Binders/BindingSourceModelBinder.cs:
```diff
@@ -59,6 +59,10 @@ protected BindingSourceModelBinder([NotNull] BindingSource bindingSource)
+        /// <remarks>
+        /// Other model binders will never run if this method is called. Return <c>null</c> to skip other model binders
+        /// but allow higher-level handling e.g. falling back to empty prefix.
+        /// </remarks>
```
This isn't true - when you return null other model binders will still run. 

What you're likely seeing is that our model binders are exclusive. If BodyModelBinder doesn't handle something nothing else will pick it up. However if you did this with something like ByteArrayModelBinder the collection model binder would at least _try_.

dougbu replied:
no, I'm saying this specific `abstract` method leaves the `override` no way to let other binders run.  in your example, the `BodyModelBinder` implementation won't call this method.  it handles exclusivity before calling `BindModelCoreAsync`.

rynowak replied:
I see.. You're turning a `null` return from `BindModelCoreAsync` into an `isModelSet == true` value. That seems fine. I think somehow I missed this important detail sorry.

dougbu replied:
nit: not my behaviour. I just documented something that was `true` before I started. :smiley_cat:

rynowak commented on test/Microsoft.AspNet.Mvc.ModelBinding.Test/Binders/KeyValuePairModelBinderTest.cs:
```diff
@@ -28,8 +26,7 @@ public class KeyValuePairModelBinderTest
             var retVal = await binder.BindModelAsync(bindingContext);
```
can we pick something better than `retVal` here? I know it's not your fault :laughing:

dougbu replied:
:ok:

rynowak commented on test/Microsoft.AspNet.Mvc.ModelBinding.Test/Binders/KeyValuePairModelBinderTest.cs:
```diff
@@ -28,8 +26,7 @@ public class KeyValuePairModelBinderTest
             // Assert
-            Assert.NotNull(retVal);
-            Assert.Null(retVal.Model);
+            Assert.Null(retVal);
```
I don't get why this case is a `null` with errors. We've found part of the model - at this point we've commited to binding using whatever prefix is currently in use.

dougbu replied:
the only non-comment change I made to `KeyValuePairModelBinder` was to remove extra special case returns, leaving that to the existing handling in `GenericModelBinder`.

so, this isn't a `null` with errors that will ever be visible.  just looks `null` temporarily because we're testing `KeyValuePairModelBinder` instead of its `GenericModelBinder` caller.  `GenericModelBinder` never returns `null` if it can find a binder.

rynowak replied:
> the only non-comment change I made to KeyValuePairModelBinder was to remove extra special case returns, leaving that to the existing handling in GenericModelBinder

If it's not visible why not just return a value then? I don't get why you choose to do it differently when it doesn't affect the outcome of the system. It's much more conceptually symmetric to return a value - that indicates that the KVPMB did *indeed* do something. Returning `null` just requires extra explanation even if it does result in the right outcome.

dougbu replied:
I made a few changes (2 or 3 commits) as I went through trying to understand the system.  this was part of that -- the multiple locations where `isModelSet: false` was used but just didn't matter was getting in the way and made the system less understandable.

if you don't like it, I can revert it.  as I said [elsewhere](https://github.com/aspnet/Mvc/pull/2188#discussion_r26820011), that'll take some additional time (hour or so) and I'm done tonight almost.

rynowak commented on test/Microsoft.AspNet.Mvc.ModelBinding.Test/Binders/MutableObjectModelBinderTest.cs:
```diff
@@ -605,6 +604,7 @@ public void GetMetadataForProperties_WithoutBindAttribute()
+                nameof(Person.ValueTypeRequiredWithDefaultValue),
                 "FirstName",
                 "LastName",
                 "PropertyWithDefaultValue",
```
If you're going to use `nameof(...)` then just make it uniform

rynowak commented on test/WebSites/FiltersWebSite/Controllers/ActionFilterController.cs:
```diff
@@ -37,6 +37,7 @@ public override void OnActionExecuting(ActionExecutingContext context)
             {
                 filters = (List<ContentResult>)obj;
             }
+            else
```
???

dougbu replied:
have no idea how the test was passing.  this is another separate commit to keep the quick fix obvious.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ModelBinders/BodyModelBinder.cs:
```diff
@@ -57,7 +55,9 @@ public class BodyModelBinder : BindingSourceModelBinder
-                return new ModelBindingResult(null, bindingContext.ModelName, isModelSet: false);
+
+                // Base class will create a ModelBindingResult that exits current ModelBinding loop.
+                return null;
```
Why is this changing? 

It seems like it has no tangible effect since the base-class just does what the old code would do.

dougbu replied:
as I mentioned in the PR description, it helps to look at the commit descriptions.  you're looking at part of one of the earliest commits -- the second.  in that I "reduce special-case `ModelBindingResult` creation locations (return `null`)"  to make the special-cases more visible.

unfortunately I also named parameters and improved some comments in that commit.  but if you hate the change, I can back it out or back parts of it out.

rynowak replied:
I don't agree at all that these are special cases. The contract of a model binder is that you return a `ModelBinderResult` if you *handled it* and `null` if you did not.  

Handling it might mean that you failed to bind, and we provide a way to indicate to the world (`isModelSet = false`). Doing things differently in two places just requires more explanation and comments.

dougbu replied:
@rynowak please stop discussing why you dislike the change and make a call on whether it's worth reverting.

rynowak replied:
If it's my call then I think it's worth reverting.

dougbu replied:
:ok:

rynowak commented:
There's no raisin bran emoji, or I would use it here. If you're not convinced about the `null` thing lets dicuss that and the model state piece tomorrow first thing. I think in terms of cleanliness and test coverage otherwise this is ready to go.

dougbu commented:
rebased and added more tests; have removed "[Design]" tag from title
- comment changes and line-level reverts remain for tomorrow morning

dougbu commented:
done and dusted I hope; two new commits covering PR comments

rynowak commented:
:shipit:

harshgMSFT commented:
:shipit:

dougbu commented:
commits 103538b thru 533474d
